prompt2model/param_selector/search_with_optuna.py

```diff
@@ -0,0 +1,245 @@
+        )
+
+        model.save_pretrained(best_model_path)
+        tokenizer.save_pretrained(best_model_path)
+        return model, tokenizer
+
+    def _build_hp_space(
+        self, hyperparameter_space: Optional[dict[str, Any]] = None
+    ) -> dict[str, Any]:
+        return (
+            {
+                "min_weight_decay": hyperparameter_space.get(
+                    "min_weight_decay",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("min_weight_decay"),
+                ),
+                "max_weight_decay": hyperparameter_space.get(
+                    "max_weight_decay",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("max_weight_decay"),
+                ),
+                "min_learning_rate": hyperparameter_space.get(
+                    "min_learning_rate",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("min_learning_rate"),
+                ),
+                "max_learning_rate": hyperparameter_space.get(
+                    "max_learning_rate",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("max_learning_rate"),
+                ),
+                "min_num_train_epochs": hyperparameter_space.get(
+                    "min_num_train_epochs",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("min_num_train_epochs"),
+                ),
+                "max_num_train_epochs": hyperparameter_space.get(
+                    "max_num_train_epochs",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("max_num_train_epochs"),
+                ),
+                "save_strategy": hyperparameter_space.get(
+                    "save_strategy",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("save_strategy"),
+                ),
+                "evaluation_strategy": hyperparameter_space.get(
+                    "evaluation_strategy",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("evaluation_strategy"),
+                ),
+                "per_device_train_batch_size": hyperparameter_space.get(
+                    "per_device_train_batch_size",
+                    DEFAULT_HYPERPARAMETERS_SPACE.get("per_device_train_batch_size"),
+                ),
+            }
+            if hyperparameter_space is not None
+            else DEFAULT_HYPERPARAMETERS_SPACE
+        )
+
+    def select_from_spec(
+        self, training_sets: list[Dataset], validation: Dataset, prompt_spec: PromptSpec
+    ) -> tuple[PreTrainedModel, PreTrainedTokenizer]:
+        """Select a model among a set of hyperparameters (given or inferred).
+
+        Args:
+            training_sets: One or more training datasets for the trainer.
+            validation: A dataset for computing validation metrics.
+            prompt_spec: A prompt to infer hyperparameters from.
+
+        Return:
+            A model and tokenizer (with hyperparameters from inferred range).
+        """
+        return super().select_from_spec(training_sets, validation, prompt_spec)
```

Comment:
Pretty sure this function isn't doing anything, since it's already inherited from the `ParamSelector` superclass

Reply:
Yes, it is not doing anything, should I remove this?

Reply:
Huh, what was the error during pre-commit? I'm wondering if the error is that this function is never used.

If that's the case, then you can leave it here as you've done, and I can purge this method in a future PR.

Reply:
Yeah will discuss that on discord.

Reply:
Done, removed this function and there is no problem.